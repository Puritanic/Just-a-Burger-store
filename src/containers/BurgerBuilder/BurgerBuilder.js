import React, { Component } from 'react';
import { connect } from 'react-redux';
import Aux from '../../hoc/Aux/Aux';
import Burger from '../../components/Burger/Burger';
import BuildControls from '../../components/Burger/BuildControls/BuildControls';
import Modal from '../../components/UI/Modal/Modal';
import OrderSummary from '../../components/Burger/OrderSummary/OrderSummary';
import Spinner from '../../components/UI/Spinner/Spinner';
import withErrorHandler from '../../hoc/withErrorHandler/withErrorHandler';
import axios from '../../axios-orders';
import {
  addIngredient,
  removeIngredient,
  initIngredients
} from '../../store/actions/burgerActions';
import { purchaseLoad } from '../../store/actions/orderActions';
import { setAuthRedirect } from '../../store/actions/auth';

export class BurgerBuilder extends Component {
  // constructor(props) {
  //     super(props);
  //     this.state = {...}
  // } or like this in React 16:
  state = {
    purchasing: false
  };

  componentDidMount() {
    // console.log(this.props);
    this.props.initIngredients();
  }

  updatePurchaseState = (ingredients) => {
    const sum = Object.keys(ingredients)
      .map(igKey => ingredients[igKey])
      .reduce((prev, curr) => prev + curr, 0);

    return sum > 0;
  };

  purchaseHandler = () => {
    if (this.props.authenticated) {
      this.setState({ purchasing: true });
    } else {
      this.props.setAuthRedirect('/checkout');
      this.props.history.push('/auth');
    }
  };

  purchaseCancelHandler = () => {
    this.setState({ purchasing: false });
  };

  purchaseContinueHandler = () => {
    this.props.purchaseLoad();
    this.props.history.push('/checkout');
  };

  render() {
    const disabledInfo = {
      ...this.props.ingredients
    };
    // eslint-disable-next-line
    for (const key in disabledInfo) {
      disabledInfo[key] = disabledInfo[key] <= 0;
    }
    let orderSummary = null;
    let burger = this.props.error ? (
      <p>Ingredients can&quot;t be loaded!</p>
    ) : (
      <Spinner />
    );

    if (this.props.ingredients) {
      burger = (
        <Aux>
          <Burger ingredients={this.props.ingredients} />
          <BuildControls
            ingredientAdded={this.props.addIngredient}
            ingredientRemoved={this.props.removeIngredient}
            disabled={disabledInfo}
            isAuth={this.props.authenticated}
            purchasable={this.updatePurchaseState(this.props.ingredients)}
            ordered={this.purchaseHandler}
            price={this.props.totalPrice}
          />
        </Aux>
      );
      orderSummary = (
        <OrderSummary
          ingredients={this.props.ingredients}
          price={this.props.totalPrice}
          purchaseCancelled={this.purchaseCancelHandler}
          purchaseContinued={this.purchaseContinueHandler}
        />
      );
    }
    if (this.props.loading) {
      orderSummary = <Spinner />;
    }
    // {salad: true, meat: false, ...}
    return (
      <Aux>
        <Modal
          show={this.state.purchasing}
          modalClosed={this.purchaseCancelHandler}
        >
          {orderSummary}
        </Modal>
        {burger}
      </Aux>
    );
  }
}

const mapStateToProps = state => ({
  ingredients: state.burgerBuilder.ingredients,
  totalPrice: state.burgerBuilder.totalPrice,
  error: state.burgerBuilder.error,
  authenticated: !!state.auth.token
});

const mapDispatchToProps = dispatch => ({
  addIngredient: ingName => dispatch(addIngredient(ingName)),
  removeIngredient: ingName => dispatch(removeIngredient(ingName)),
  initIngredients: () => dispatch(initIngredients()),
  purchaseLoad: () => dispatch(purchaseLoad()),
  setAuthRedirect: path => dispatch(setAuthRedirect(path))
});

export default connect(mapStateToProps, mapDispatchToProps)(withErrorHandler(BurgerBuilder, axios));
