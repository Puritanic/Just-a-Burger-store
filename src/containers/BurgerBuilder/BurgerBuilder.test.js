import { shallow, configure } from 'enzyme';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { BurgerBuilder } from './BurgerBuilder';
import BuildControls from '../../components/Burger/BuildControls/BuildControls';

configure({ adapter: new Adapter() });

describe('<BurgerBuilder /> container', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<BurgerBuilder initIngredients={() => {}} />);
  });

  it('should render Burger Builder container correctly', () => {
    wrapper.setProps({ ingredients: { salad: 0 } });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render BuildControls in BurgerBuilder container correctly', () => {
    wrapper.setProps({ ingredients: { salad: 0 } });
    expect(wrapper.find(BuildControls)).toHaveLength(1);
  });
});
