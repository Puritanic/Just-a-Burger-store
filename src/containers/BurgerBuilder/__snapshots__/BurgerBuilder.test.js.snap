// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BurgerBuilder /> container should render Burger Builder container correctly 1`] = `
<aux>
  <Modal
    modalClosed={[Function]}
    show={false}
  >
    <OrderSummary
      ingredients={
        Object {
          "salad": 0,
        }
      }
      purchaseCancelled={[Function]}
      purchaseContinued={[Function]}
    />
  </Modal>
  <aux>
    <burger
      ingredients={
        Object {
          "salad": 0,
        }
      }
    />
    <buildControls
      disabled={
        Object {
          "salad": true,
        }
      }
      ordered={[Function]}
      purchasable={false}
    />
  </aux>
</aux>
`;
